### ISSUE TITLE
Support for dual SIM phones

### ISSUE THREAD

**Issue by Kiwawa** [SecUpwN/Android-IMSI-Catcher-Detector#504]
I have a Phablet with dual SIMs. One 2G, one 3G, according to the manufacturers specs. Default for all services is set to second, 3G-enabled, SIM ("Comviq 2") as seen in the screenshot. The application only detects / displays information about the first SIM.

**Comment by SecUpwN** [comment]
Good evening @Kiwawa. Our app apparently does not support Dual-SIM phones.
Security notice: You should obfuscated LAC/CID from your above screenshot.

**Comment by Kiwawa** [comment]
Hi and thanks for the note. I don't mind giving away my approx. location, but removed the screenshots anyway. Perhaps you can disregard this bug report then, or turn it into an improvement suggestion? Happy to help...

**Comment by SecUpwN** [comment]
>Perhaps you can disregard this bug report then, or turn it into an improvement suggestion?

Re-labeled this Issue and will wait for the others to respond. If we ever support Dual SIM phones, it will not be right away since we clearly lack contributors and need to focus on fixing existing bugs and continue to add the core functions of our app. Feel invited to spread the word to Android developers!

**Comment by TPS** [comment]
Just as a note, this DOES seem to overlap a little with #457, as some of the work that'll be done there can likely be expanded toward this issue, I think. ☺

**Comment by ehem** [comment]
I must disagree with SecUpwN's removal of the BUG tag.  Dual-SIM (even triple-SIM!) phones are very real and not working on real hardware is a bad thing.  The additional data could be interesting since if one SIM sees an extra "cell tower" while the other doesn't might be a hint that the extra "tower" is in fact a stingray.

I'd like to help and be looking into how the Android APIs work, but I've already got too many things on my plate (may happen in the future, but not yet).

**Comment by SecUpwN** [comment]
>I must disagree with SecUpwN's removal of the BUG tag.

Disagree all you want, @ehem, but this is *not* a "bug", it is an *enhancement* if we ever add it.

> Dual-SIM (even triple-SIM!) phones are very real and not working on real hardware is a bad thing.

Our app *is* working on these phones, it just does not show and dissect the data of the other SIMs. We simply have more important Issues to solve at this moment, we'd be happy to have more developers working on our app to even achieve our [basic app goals](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/wiki/Development-Status#application-goals) first. So please be patient, would you?

>The additional data could be interesting since if one SIM sees an extra "cell tower" while the other doesn't might be a hint that the extra "tower" is in fact a stingray.

Thanks for the suggestion, but do you have any documentation that this assumption must be true?

>I'd like to help and be looking into how the Android APIs work, but I've already got too many things on my plate (may happen in the future, but not yet).

My fingers are itching to be sarcastic and say "what a suprise", but sadly it really is none.

**Comment by ehem** [comment]
> >I'd like to help and be looking into how the Android APIs work, but I've already got too many things on my plate (may happen in the future, but not yet).
> 
> My fingers are itching to be sarcastic and say "what a suprise", but sadly it really is none.

I can't blame you for thinking that.  I do hope to look at things further, but there are other open source projects that I'm involved with as well.  Within a few days hopefully one minor *WRT distribution will be getting an updated set of kernel patches.  Then I've got some Android lock screen experimentation to do, then I may be looking back at AIMSICD.


Some notes on dual-SIM devices.  I believe *most* dual-SIM devices allow one SIM to operate in 2G/3G/4G while the other is 2G-only.  The one operating in 3G/4G mode can be easily switched, but only one at a time acts in 3G/4G and this one is also used for data.

Appears on my device AIMSICD is reportting data from the first SIM and ignoring the data from the second SIM/radio. If I tell it to turn off the first SIM, the data under "SIM Information" is blank.  If I *completely* remove the first SIM, then AIMSICD will report data from the second SIM slot (after rebooting, since removing the SIM requires powering off the device).

**Comment by SecUpwN** [comment]
>I can't blame you for thinking that.

Rest assured that I'm not blaming *you* in specific, but generally how few people actually contribute given the efforts of in how many places I continuously keep introducing our app to people, websites, etc. Please take all the time you need to finish the other things you're aiming to look at. We're *very* happy if you come back once in a while and actually use our app and note down improvements.

Regarding this Issue, I'm not quite sure how you'd map out an "ideal solution". We currently only have one main screen and I'm sure you'd like to see the data of the second and third SIM as well? Tell us.

**Comment by E3V3A** [comment]
Do I need to lock this issue or block the user?

**Comment by SecUpwN** [comment]
>Do I need to lock this issue or block the user?

Neither of these options. @ehem is a good guy. He understood.

**Comment by SecUpwN** [comment]
@E3V3A, I do not want to step on your toes, but closing Issues without a proper explanation to the user who filed it will not bring freedom to our user base. I have specifically added the label `DEV_DISCUSSION` to **discuss**  this to to find out: Are we going to support dual SIM / triple SIM phones at all? @He3556 and @banjaxbanjo, is there even the *possibility* that our app could ever check two or three SIMs at the same time as a measure to detect a rogue base station? Do we need to tell our users "Folks, our app is aiming to work on as many **single SIM** phones as possible"?

**Comment by DJaeger** [comment]
I think we should at least support multiple SIMs in a basic behavior.
The should at least use/check/track the towerdata of both SIMSs.
As they can be of different providers they will see different cells.

Also the SIM details should be show in the SIM details activity.
This could be done with a dropdown to choose one of them or by simply view them one under the other.

**Comment by E3V3A** [comment]
> The should at least use/check/track the towerdata of both SIMSs
> As they can be of different providers they will see different cells.

I don't see any benefit at all for being able to do this. This OT "feature" seem totally irrelevant for the primary goal of AIMSICD. Use your SIM manager to switch your SIM or physically swap them. 

The **only** possible benefit for this, is **IF** you (for example) could access the baseband and force SIM-1 to use LTE and SIM-2 to use UMTS, on the **same** MNC. I don't see anyone presently here, to have to skills to implement this for the foreseeable future. Which is why I closed this issue as invalid.

**Comment by DJaeger** [comment]
The primary goal of AIMSICD is to detect IMSI Catchers.
Why should we only detect them on the first SIM?
What is with IMSI catchers the second SIM is connecting to?

**Comment by SecUpwN** [comment]
Thank you for adding your valuable opinions, folks. I *really* like what has been proposed in https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/issues/504#issuecomment-121856461 and am voting for the *dropdown menu* so that details for each SIM can be compared by the user. @DJaeger, are you able to tackle this first basic implementation? Any further enhancements to this are appreciated as well.

**Comment by E3V3A** [comment]
@SecUpwN Seem like you're not believing what I'm saying. I repeat, I don't see anyone capable of modifying the baseband behavior here, so, drop it! 

If you just wanna see the IMEI, and service provider for both, fine. But don't think for a second you're gonna be able to use both at the same time in this app.

**Comment by SecUpwN** [comment]
@E3V3A, so you don't think that we should at least implement some basic checks?

>We should at least use/check/track the towerdata of both SIMSs.
>As they can be of different providers they will see different cells.

**Comment by ehem** [comment]
> The **only** possible benefit for this, is **IF** you (for example) could access the baseband and force SIM-1 to use LTE and SIM-2 to use UMTS, on the **same** MNC. I don't see anyone presently here, to have the skills to implement this for the foreseeable future. Which is why I closed this issue as invalid.

While I've got no idea whether this cannot be controlled by AIMSICD, precisely *this* is entirely possible with my phone. The 3G/4G protocols can only be used on one SIM at a time, the other SIM can _only_ be using the 2G protocol.  The SIM which the 2G/3G/4G hardware is talking to can be forced to _only_ use the 3G/4G protocols.  While I doubt AIMSICD could force both SIMs to use the same MNC, AIMSICD can check whether both SIMs are talking to the same MNC and which protocol version they're using at which point we've met the requirements for the above.

I was originally thinking more along the lines of what @DJaeger was thinking. Merely monitoring both SIMs may provide more data.

If entities are only targetting one SIM, suspicious behavior _may_ be more readily detected on the other SIM.  Heavy differences between SIMs from one carrier may be suspicious (though this may not be detectable).  Two SIMs from distinct carriers converging on the same tower, particularly if "native" towers are visible may also be suspicious.

**Comment by DJaeger** [comment]
Why should it be suspicious if a phone with SIMs from one carrier are connected to different cells?
Is it not possible on moving, that one has already connected to the next cell, while one is still connected to the previous and will change a little bit later?

If both SIMs are from different carriers and see the same cell this is realy suspicious.
But is this possible, as they should have different MNCs?
Doesn't Android already prevent this?

**Comment by DJaeger** [comment]
What is the problem with "may provide more data." ?
This is the part I mean, what we should do.

**Comment by SecUpwN** [comment]
>If both SIMs are from different carriers and see the same cell this is realy suspicious.
>But is this possible, as they should have different MNCs?
>Doesn't Android already prevent this?
>What is the problem with "may provide more data." ?

@E3V3A and @He3556, please help to answer the quoted questions. Thank you.

**Comment by johanw666** [comment]
From what I understand from discussions in other projects, Android before 5.1 has no standard way to deal with multi-sim phones; it is put in by the builders of those phones. So to support this in those phones requires a vendor-specific solution, and there are a lot of obscure vendors buidlign multi-sim phones so supporting them would be a lot of work.

**Comment by TPS** [comment]
FWIW, #898 contains docs with some Stingray details.

**Comment by ivbeg** [comment]
Dual sim phones are standard de-facto in Russia now. 
Is it possible to add at least sim-switching ? 
How I could support fixing this issue? Bounty source https://www.bountysource.com or something like that  ?
